### PR TITLE
[wip] Rackspace Provider

### PR DESCRIPTION
**Pull request by jrperritt**
This needs more acceptance tests and documentation before it's ready to merge. Unlike with the OpenStack provider, I'll just get these resources merged and then add on others in other PRs.
- [ ] acceptance tests
- [ ] documentation

### CODE REVIEW

**Comment by mitchellh** [comment]
Just curious: is RackSpace not OpenStack compatible? We support Openstack now, so will that not work?

**Comment by jrperritt** [comment]
It would mostly work, but there are some differences. Neutron (networking), for example, is the same. But some services offer additional or fewer request options (for example, when creating a server).

**Comment by nmarasoiu** [comment]
Indeed, I think authentication differs too, because I tried a lot of combinations and I could not log in to rackspace auth with openstack connector of either terraform, docker-machine or openstack command line, and I only could connect with nova client packaged by rackspace and presumably bundling their plugin with the differences. Very dissapointing.

Do you have an idea when this Rackspace connector / module will be available? 
Thank you

**Comment by jrperritt** [comment]
Yes, authentication is quite incompatible. I'd started this months ago, but other priorities have since taken precedent; I'm not sure when I'll be able to get back to this. What's there now still needs tests, testing and documentation. If you like, you can make PRs to my branch.

**Comment by nmarasoiu** [comment]
Hi,
The branch, 801d0c5c4473429fa80dfc489fbd6b9cb97eda8c , gives me:
FAIL	github.com/hashicorp/terraform/command [build failed]
Is this correct?

**Comment by nmarasoiu** [comment]
And also:
make dev
go generate ./...
==> Getting dependencies...
# github.com/hashicorp/terraform/command
command/meta.go:351: cannot use ColorizeUi literal (type *ColorizeUi) as type cli.Ui in field value:
	*ColorizeUi does not implement cli.Ui (missing AskSecret method)
make: *** [dev] Error 2

**Comment by nmarasoiu** [comment]
the rackspace guys are saying default openstack tooling should work if you pass your password rather than api key and set NOVA_SERVICE_NAME=cloudServersOpenStack but they admit working on true compatibility:)

File: builtin/providers/rackspace/config.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	"regexp"
+	"runtime"
+	"strings"
+
+	"github.com/rackspace/gophercloud"
+	"github.com/rackspace/gophercloud/rackspace"
+)
+
+// Config represents the options a user can provide to authenticate to a
+// Rackspace endpoing
```
Comment (rickhull) [inline builtin/providers/rackspace/config.go]: *endpoint

**Comment by st33v** [comment]
:+1:

I could really use this right about now.

**Comment by nmarasoiu** [comment]
Hi,

I could use a terraform openstack script to ramp up a cluster, by setting
an additional env var that instructed rackspace to be openstack compatible.
pls find more on this thread up or rackspace irc and try it and confirm us
if it is compatible like this

În data de luni, 13 iulie 2015, Steve Neuharth <notifications@github.com> a
scris:

**Comment by rphillips** [comment]
@nmarasoiu can you post what you tried to this PR? It does not look like it got saved to the github website.

https://github.com/hashicorp/terraform/pull/1438

**Comment by nmarasoiu** [comment]
Hi,

By giving OS_PASSWORD the password (not the api key), or the other way around; and by adding an extra OS_ variable which I do not remember right now, I was able to create a cluster with terraform on rackspace using the openstack module.

If you don;t find that extra variable, please let me know. I found it asking on the freenode channel of rackspace. I changed jobs and I don't have the information at hand anymore.

Good luck!

**Comment by joemiller** [comment]
Perhaps these vars will work? I have not used these with terraform yet, but
I use them with packer which in turn uses the gophercloud lib, and perhaps
terraform does too.

export OS_USERNAME=user
export OS_PASSWORD=pass
export OS_TENANT_NAME=<get tenant-id from the url while logged into
mycloud.rackspace.com>
export OS_AUTH_URL=https://identity.api.rackspacecloud.com/v2.0/

**Comment by rphillips** [comment]
Thanks everyone. Looks like this is working with the openstack provider for cloud instances. Networks still has that bug in gophercloud appending two `/v2.0/v2.0` for the url.

**Comment by koen-serry** [comment]
Unsure if we're ever see this pull request in the master, but I just tried with the openstack provider and I've got issues (http status 404 on both security groups and networking)

Error creating OpenStack security group: Expected HTTP response code [200] when accessing [POST https://lon.servers.api.rackspacecloud.com/v2/10046442/os-security-groups], but got 404 instead

Error creating OpenStack Neutron network: Expected HTTP response code [201 202] when accessing [POST https://lon.networks.api.rackspacecloud.com/v2.0/v2.0/networks], but got 404 instead

**Comment by josephholsten** [comment]
@jrperritt are you still working on this?

**Comment by jrperritt** [comment]
Not in the foreseeable future. That is, priorities have since shifted, and this is not currently on the list. It wouldn't be difficult for someone to get this PR to a mergeable state. It needs testing and docs (and maybe some tweaks in response to the testing).

**Comment by phinze** [comment]
Thanks for all the work on this so far, @jrperritt. Going to close since it's not actively being worked on, but we're definitely willing to help out anybody who'd like to pick it back up to get it over the finish line.

**Comment by gbjk** [comment]
I've just finished writing a provider for bigv. We also need one for rackspace. We'll be picking this up and running with it so far as our use-case demands ( create, modify, destroy vms, networking, etc ).
Probably start off with it in a new repo, and go from there.

Just wanted to comment so there's hope for anyone feeling depressed about this.

**Pull request closed without merging**